File: server/src/main/java/au/com/codeka/warworlds/server/html/render/StarRendererHandler.java
package au.com.codeka.warworlds.server.html.render;

import au.com.codeka.warworlds.common.Log;
import au.com.codeka.warworlds.common.proto.Star;
import au.com.codeka.warworlds.server.handlers.RequestException;
import au.com.codeka.warworlds.server.world.StarManager;
import au.com.codeka.warworlds.server.world.WatchableObject;
import java.io.File;
import java.util.Locale;
import java.util.Random;

/**
 * {@link RendererHandler} for rendering stars.
 */

public class StarRendererHandler extends RendererHandler {
  private static final Log log = new Log("StarRendererHandler");

  @Override
  public void get() throws RequestException {
    long starId = Long.parseLong(getUrlParameter("star"));
    int width = Integer.parseInt(getUrlParameter("width"));
    int height = Integer.parseInt(getUrlParameter("height"));
    String bucket = getUrlParameter("bucket");
    Float factor = BUCKET_FACTORS.get(bucket);
    if (factor == null) {
      log.warning("Invalid bucket: %s", getRequest().getPathInfo());
      getResponse().setStatus(404);
      return;
    }

    WatchableObject<Star> star = StarManager.i.getStar(starId);
    File cacheFile = new File(String.format(Locale.ENGLISH,
        "data/cache/star/%d/%dx%d/%s.png", starId, width, height, bucket));
    if (cacheFile.exists()) {
      serveCachedFile(cacheFile);
      return;
    } else {
      cacheFile.getParentFile().mkdirs();
    }

    Random rand = new Random(starId);
    File templateFile = getTemplateFile(rand, "star", star.get().classification.toString());
    if (templateFile == null) {
      getResponse().setStatus(500);
      return;
    }

    long startTime = System.nanoTime();
    if (!generateImage(cacheFile, templateFile, null, width, height, factor, rand)) {
      getResponse().setStatus(500);
      return;
    }
    long endTime = System.nanoTime();
    log.info("%dms to generate image for %s",
        (endTime - startTime) / 1000000L, getRequest().getPathInfo());

    serveCachedFile(cacheFile);
  }

}

File: server/src/main/kotlin/au/com/codeka/warworlds/server/html/render/StarRendererHandler.kt
package au.com.codeka.warworlds.server.html.render

import au.com.codeka.warworlds.common.Log
import au.com.codeka.warworlds.server.handlers.RequestException
import au.com.codeka.warworlds.server.world.StarManager
import java.io.File
import java.util.*

/** [RendererHandler] for rendering stars. */
class StarRendererHandler : RendererHandler() {
  private val log = Log("StarRendererHandler")

  public override fun get() {
    val starId = getUrlParameter("star")?.toLong() ?: throw RequestException(404, "Invalid star")
    val width = getUrlParameter("width")?.toInt() ?: throw RequestException(404, "Invalid width")
    val height = getUrlParameter("height")?.toInt() ?: throw RequestException(404, "Invalid height")
    val bucket = getUrlParameter("bucket")
    val factor = BUCKET_FACTORS[bucket] ?: throw RequestException(404, "Invalid bucket")

    val cacheFile = File("data/cache/star/${starId}/${width}x${height}/${bucket}.png")
    if (cacheFile.exists()) {
      serveCachedFile(cacheFile)
      return
    }
    cacheFile.parentFile.mkdirs()

    val star = StarManager.i.getStarOrError(starId)
    val rand = Random(starId)
    val templateFile = getTemplateFile(rand, "star", star.get().classification.toString())
    val startTime = System.nanoTime()
    generateImage(cacheFile, templateFile, null, width, height, factor, rand)
    val endTime = System.nanoTime()
    log.info("%dms to generate image for %s", (endTime - startTime) / 1000000L, request.pathInfo)
    serveCachedFile(cacheFile)
  }

}